rootmap/sqlrootmap.go

```diff
@@ -65,23 +201,48 @@ func (rm *sqlRootmap) Put(
 ) error {
 	rm.mtx.Lock()
 	defer rm.mtx.Unlock()
-	_, err := rm.db.ExecContext(ctx, `
-	insert into rootmap (
-		database,
-		producer_id,
-		topic,
-		version,
-		storage_prefix,
-		node_id
-	) values ($1, $2, $3, $4, $5, $6)`,
-		database, producerID, topic, version, prefix, hex.EncodeToString(nodeID[:]),
-	)
-	if err != nil {
-		var err sqlite3.Error
-		if errors.As(err, &sqlite3.Error{Code: sqlite3.ErrConstraint}) {
-			return ErrRootAlreadyExists
+	err := rm.withTransaction(func(tx *sql.Tx) error {
+		var tableID int
+		err := rm.db.QueryRowContext(ctx, `
+		select id from tables where producer = ? and topic = ? and database = ? 
+		`, producerID, topic, database).Scan(&tableID)
+		if err != nil && !errors.Is(err, sql.ErrNoRows) {
+			return fmt.Errorf("failed to read from tables: %w", err)
 		}
-		return fmt.Errorf("failed to store to rootmap: %w", err)
+		if errors.Is(err, sql.ErrNoRows) {
+			err := rm.db.QueryRowContext(ctx, `
+			insert into tables (database, producer, topic) values (?, ?, ?)
+			returning id
+			`, database, producerID, topic).Scan(&tableID)
+			if err != nil {
+				return fmt.Errorf("failed to insert into tables: %w", err)
+			}
+			rm.tables[table{database, producerID, topic}] = true
```

Comment:
once the transaction is properly hooked up, this is actually a bug. If a transaction rolls back, the cache state will still reflect that the table is present, but it will have been removed from the DB.

I think what we'll want to do is declare that tableID variable outside the tx, and only populate the cache once we have exited the transaction.

Another change I think we can make is change rm.tables from map[table]bool to map[table]int64, and then store the IDs in the values. That would let us cut out a database access in this function, possibly others elsewhere.

Reply:
Done, I believe!